### PR TITLE
Add Keithley DAQ6510

### PR DESCRIPTION
**Pull request by Aphelion82**
Add support for the [Keithley DAQ6510 Data Acquisition and Logging Multimeter System](https://www.tek.com/en/products/keithley/digital-multimeter/keithley-daq6510).

### CODE REVIEW

File: pymeasure/instruments/keithley/keithleyDAQ6510.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+    open_channel = Instrument.setting(
+        ":ROUT:OPEN (@%d)",
+        """Set a single channel to open."""
+    )
+
+    close_channel = Instrument.setting(
+        ":ROUT:CLOS (@%d)",
+        """Set a single channel to closed."""
+    )
```
Comment (BenediktBurger) [inline pymeasure/instruments/keithley/keithleyDAQ6510.py:56]: You use it as an "action word", therefore these properties should be methods instead.

**Comment by BenediktBurger** [comment]
Thanks for your contribution.

I do not get, what you can do with this device. Until now, you can only open a channel, but what does that do?

File: pymeasure/instruments/keithley/keithleyDAQ6510.py (hunk 1 above)
Comment (Aphelion82) [reply pymeasure/instruments/keithley/keithleyDAQ6510.py]: Changed to be methods

**Comment by Aphelion82** [comment]
> Thanks for your contribution.
> 
> I do not get, what you can do with this device. Until now, you can only open a channel, but what does that do?

The DAQ6510 is able to measure voltage and current, among other things. I haven't yet implemented these

**Comment by BenediktBurger** [comment]
That explains it.

**Comment by Aphelion82** [comment]
Thank you—as far as the opening of channels, the DAQ6510 uses a MUX card with channels that can be opened and closed.

The codecov check fails due to a codecov token being missing. Do I need to add that on my end?

**Comment by BenediktBurger** [comment]
> The codecov check fails due to a codecov token being missing. Do I need to add that on my end?

No, you cannot add the token. And for token less upload, codecov runs sometimes into api call limits. So, don't worry.

**Comment by Aphelion82** [comment]
The checks fail on the docstrings, but I had been following the wording used by preexisting instrument files (see [keithley2400](https://github.com/pymeasure/pymeasure/blob/606fae7e8063594394a089ee88d2aa5df8f19570/pymeasure/instruments/keithley/keithley2400.py#L151-L163) and [keithley2450](https://github.com/pymeasure/pymeasure/blob/606fae7e8063594394a089ee88d2aa5df8f19570/pymeasure/instruments/keithley/keithley2450.py#L91-L104)) for similar/identical purposes. Should I still change the docstrings, or leave it as it is?

**Comment by BenediktBurger** [comment]
> The checks fail on the docstrings, but I had been following the wording used by preexisting instrument files (see [keithley2400](https://github.com/pymeasure/pymeasure/blob/606fae7e8063594394a089ee88d2aa5df8f19570/pymeasure/instruments/keithley/keithley2400.py#L151-L163) and [keithley2450](https://github.com/pymeasure/pymeasure/blob/606fae7e8063594394a089ee88d2aa5df8f19570/pymeasure/instruments/keithley/keithley2450.py#L91-L104)) for similar/identical purposes. Should I still change the docstrings, or leave it as it is?

Please change the docstrings. We adjusted at some point in history the [coding standards](https://pymeasure.readthedocs.io/en/latest/dev/coding_standards.html#docstrings) (in order to be in line with python recommendations), however we did not change all the docstrings of all instruments yet, as it is quite some work to change the wrong docstrings of all instruments.

File: pymeasure/instruments/keithley/keithleyDAQ6510.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        The largest value is the longest time, and results in the lowest reading rate,
+        but increases the number of usable digits. """
+    )
+
+    ####################
+    # Methods        #
+    ####################
+
+    def measure_resistance(self, nplc=1, resistance=100e6, auto_range=True):
+        """ Configures the measurement of resistance.
```
Comment (BenediktBurger) [inline pymeasure/instruments/keithley/keithleyDAQ6510.py:183]: Docstrings should be (according to [PEP 257](https://peps.python.org/pep-0257/)) in imperative mode:
```suggestion
        """ Configure the measurement of resistance.
```

Equally below.

File: pymeasure/instruments/keithley/keithleyDAQ6510.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+            self.current_range = current
+        self.check_errors()
+
+    def no_errors(self):
+        """
+        Check to see if the instrument has any errors returned or not.
+
+        :return: ``True`` if there are no errors, ``False`` if there are errors.
+        """
+        return len(self.check_errors() == 0)
```
Comment (BenediktBurger) [inline pymeasure/instruments/keithley/keithleyDAQ6510.py:242]: Did you know, that `bool([]) is False`? That means, you can use `if inst.check_errors(): handle_errors()`?

That would make that method superfluous.

A general suggestion: try to word a boolean property, that it works like a question: `no_errors() is True` (or `something_enabled = True`, as that makes it easier to interpret True/False.
In such a case (like your name), you can write in the docstring: `:return: whether there are errors.`. (or maybe `:return bool: whether there are errors.`

File: tests/instruments/keithley/test_keithleyDAQ6510_with_device.py
Comment (BenediktBurger) [inline tests/instruments/keithley/test_keithleyDAQ6510_with_device.py:1]: Thanks for adding these device tests.

However, we maintainers like to have [protocol tests](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/tests.html#protocol-tests), which do not require the instrument (which we do not have).
These protocol tests allow us to verify, that the instrument continues to work, even if we change code (in the core).

If you need help, don't hesitate to ask.

**Review by BenediktBurger — CHANGES_REQUESTED** [review]
Thanks for your contribution.

Your code is already in a good shape, only a few small issues.

File: pymeasure/instruments/keithley/keithleyDAQ6510.py (hunk 2 above)
Comment (Aphelion82) [reply pymeasure/instruments/keithley/keithleyDAQ6510.py]: Changed all that I spotted

File: pymeasure/instruments/keithley/keithleyDAQ6510.py (hunk 3 above)
Comment (Aphelion82) [reply pymeasure/instruments/keithley/keithleyDAQ6510.py]: Thanks—I just removed the method.

File: tests/instruments/keithley/test_keithleyDAQ6510_with_device.py
Comment (Aphelion82) [reply tests/instruments/keithley/test_keithleyDAQ6510_with_device.py:1]: I created a protocol test for the DAQ6510 going off of some [other tests](https://github.com/pymeasure/pymeasure/blob/b688a09a5d9b832d4bf65fea0227c780b1394c05/pymeasure/instruments/keithley/keithleyDMM6500.py) and the documentation. Let me know if there's anything wrong. Thanks for the help!

Comment (BenediktBurger) [reply tests/instruments/keithley/test_keithleyDAQ6510_with_device.py:1]: Thanks for adding them.

File: tests/instruments/keithley/test_keithleyDAQ6510.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+#
+
+from pymeasure.test import expected_protocol
+from pymeasure.instruments.keithley import KeithleyDAQ6510
+
+
+def test_init():
+    with expected_protocol(
+            KeithleyDAQ6510,
+            [(b'*LANG SCPI', None)],
```
Comment (BenediktBurger) [inline tests/instruments/keithley/test_keithleyDAQ6510.py:32]: Could it be, that you removed the "lang scpi" command from the init?

At least all tests have it and they all fail, as this command is never issued (see the failing tests).

Comment (Aphelion82) [reply tests/instruments/keithley/test_keithleyDAQ6510.py:32]: Yeah, I had wondered if that was the case, looking at the tests

Comment (Aphelion82) [reply tests/instruments/keithley/test_keithleyDAQ6510.py:32]: Removed `*LANG SCPI`

File: tests/instruments/keithley/test_keithleyDAQ6510.py [hunk 5]
```diff
@@ -26,63 +26,49 @@
 from pymeasure.instruments.keithley import KeithleyDAQ6510
 
 
-def test_init():
-    with expected_protocol(
```
Comment (BenediktBurger) [inline tests/instruments/keithley/test_keithleyDAQ6510.py]: You can keep the init test (albeit with an empty comm_pair_list), if you wish.

**Review by BenediktBurger — APPROVED** [review]
One last question, afterwards I can merge your pr, if you agree.

Thanks for your contribution.

File: tests/instruments/keithley/test_keithleyDAQ6510.py (hunk 5 above)
Comment (Aphelion82) [reply tests/instruments/keithley/test_keithleyDAQ6510.py]: Do you think it's a good idea? I'll add it back anyways.

**Comment by Aphelion82** [comment]
Thanks so much for your help.

**Comment by BenediktBurger** [comment]
Thank you very much for your contribution.

**Pull request merged**